### ISSUE TITLE
Missing user warning - qt event loop should be started before launching napari

### ISSUE THREAD

**Issue by GenevieveBuckley** [napari/napari#2233]
## 🐛 Bug

Failing to use the `%gui qt` magic in ipython/jupyter before launching napari no longer shows you a handy warning message explaining 

## To Reproduce

1. `conda create -n napari-testenv python=3.7 pip`
2. `conda activate napari-testenv`
3. Start `ipython` and then import napari and show the viewer.
```python
import napari
viewer = napari.Viewer()
```

You'll see this blank, black napari window (not what we want), and there is no error message explaining that the user should try to use `%gui qt` before starting napari.

![napari-black-screen](https://user-images.githubusercontent.com/30920819/107183673-5365f800-6a33-11eb-9010-a18622a17aa7.png)

## Expected behavior

I expected a warning message to be printed to the terminal, telling the user that they should start the qt loop (with `%gui qt` or similar) before starting napari.

## Environment

 - Please copy and paste the information at napari info option in help menubar here:
```
napari: 0.2.4.dev736+g79ab358
Platform: Linux-5.8.0-41-generic-x86_64-with-debian-bullseye-sid
System: Ubuntu 20.04.2 LTS
Python: 3.7.9 | packaged by conda-forge | (default, Dec 9 2020, 21:08:20) [GCC 9.3.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.20.1
SciPy: 1.6.0
Dask: 2021.02.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 20.2.6
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen #1: resolution 1920x1080, scale 1.0
- screen #2: resolution 1920x1080, scale 1.0

Plugins:
- svg: 0.1.4
```
 - Any other relevant information:

**Comment by tlambert03** [comment]
Hi @GenevieveBuckley,

The reason for this is that it's no longer an error to _create_ a viewer without using gui_qt first (like it used to be).  But you do need to use either gui_qt or `napari.run()` (the latter is preferable) to actually show the viewer. Think of this now as similar to having to use plt.show in matplotlib.

Basically, we're phasing out gui_qt.  See https://github.com/napari/napari/pull/2056#discussion_r550924887 for more background and rationale on this new behavior. And let us know what you think

**Comment by sofroniewn** [comment]
> You'll see this blank, black napari window (not what we want), and there is no error message explaining that the user should try to use %gui qt before starting napari.

@tlambert03 but wouldn't you expect to not even see that black screen? I actually don't see it, but i do see focus leave ipython and then return. I guess that's because `Window` is created in the init?

The discussion https://github.com/napari/napari/pull/2225#issuecomment-775157902 is very relevant to I think, and I'm curious @GenevieveBuckley if you have ideas there too!!

**Comment by tlambert03** [comment]
yeah, I don't get the black screen either... so I'd need to know more about how the script is being run.

**Comment by GenevieveBuckley** [comment]
I got the black screen on both Ubuntu 20.04 and on my mac laptop, following the "steps to reproduce" in my original comment.

I'm not sure what you mean by "how the script is being run" @tlambert03. I haven't been trying to run a python script, I've just been trying to launch napari from an interactive ipython console.

**Comment by sofroniewn** [comment]
hmm, can you try again with the latest master @GenevieveBuckley? maybe we've fixed at least this part lately

**Comment by GenevieveBuckley** [comment]
I did try with master yesterday, I'll give it another go now

**Comment by GenevieveBuckley** [comment]
Ok, here's what's happening.

I open a terminal, activate my napari-dev conda environment, and start ipython.
This works as expected:
```python
%gui qt

import napari

viewer = napari.Viewer()
# As expected, a new napari window is opened
```

But if you forget to use `%gui qt` or similar, you no longer see the helpful error message suggesting how to fix this.
```python
import napari

viewer = napari.Viewer()  # this shows the blank, black window (see screenshot in original post)
viewer.show()  # this does nothing
napari.run()  # this now shows napari in what was the blank window
# but now the ipython console is locked and won't let you type anything else
```

```

**Comment by tlambert03** [comment]
This is very strange.  Without the ipython magic and without napari.run, there should be no event loop to show any window at all.

I'd say to try a new environment but it sounds like you've already seen this on two computers.  Can you give me the full output of your pip freeze?

**Comment by tlambert03** [comment]
And what you observed above is the expected behavior, with the exception of the black screen

**Comment by GenevieveBuckley** [comment]
Yeah, the black screen is weird.

pip freeze output, napari development environment, Ubuntu 20.04
```
alabaster==0.7.12
appdirs==1.4.4
asciitree==0.3.3
attrs==20.3.0
Babel==2.9.0
backcall==0.2.0
black==20.8b1
build==0.2.0
cachey==0.2.1
certifi==2020.12.5
cfgv==3.2.0
chardet==4.0.0
check-manifest==0.46
click==7.1.2
cycler==0.10.0
dask==2021.2.0
decorator==4.4.2
distlib==0.3.1
docstring-parser==0.7.3
docutils==0.16
fasteners==0.16
filelock==3.0.12
flake8==3.8.4
freetype-py==2.2.0
fsspec==0.8.5
HeapDict==1.0.1
identify==1.5.13
idna==2.10
imageio==2.9.0
imagesize==1.2.0
importlib-metadata==3.4.0
iniconfig==1.1.1
ipykernel==5.4.3
ipython==7.20.0
ipython-genutils==0.2.0
jedi==0.18.0
Jinja2==2.11.3
jupyter-client==6.1.11
jupyter-core==4.7.1
kiwisolver==1.3.1
magicgui==0.2.6
MarkupSafe==1.1.1
matplotlib==3.3.4
mccabe==0.6.1
mypy-extensions==0.4.3
-e git+git@github.com:GenevieveBuckley/napari.git@133ab7f70803b7db26f772335b30fb96e4a44b02#egg=napari
napari-plugin-engine==0.1.9
napari-svg==0.1.4
networkx==2.5
nodeenv==1.5.0
numcodecs==0.7.3
numpy==1.20.1
numpydoc==1.1.0
packaging==20.9
pandas==1.2.1
parso==0.8.1
pathspec==0.8.1
pep517==0.9.1
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.1.0
pluggy==0.13.1
pooch==1.3.0
pre-commit==2.10.1
prompt-toolkit==3.0.14
psutil==5.8.0
ptyprocess==0.7.0
py==1.10.0
pycodestyle==2.6.0
pydantic==1.7.3
pyflakes==2.2.0
Pygments==2.7.4
PyOpenGL==3.1.5
pyparsing==2.4.7
PyQt5==5.15.2
PyQt5-sip==12.8.1
pytest==6.2.2
pytest-faulthandler==2.0.1
pytest-qt==3.3.0
pytest-timeout==1.4.2
python-dateutil==2.8.1
pytz==2021.1
PyWavelets==1.1.1
PyYAML==5.4.1
pyzmq==22.0.2
qtconsole==5.0.2
QtPy==1.9.0
regex==2020.11.13
requests==2.25.1
scikit-image==0.18.1
scipy==1.6.0
six==1.15.0
snowballstemmer==2.1.0
Sphinx==3.4.3
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==1.0.3
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.4
tifffile==2021.2.1
toml==0.10.2
toolz==0.11.1
tornado==6.1
traitlets==5.0.5
typed-ast==1.4.2
typing-extensions==3.7.4.3
urllib3==1.26.3
virtualenv==20.4.2
vispy==0.6.6
wcwidth==0.2.5
wrapt==1.12.1
xarray==0.16.2
zarr==2.6.1
zipp==3.4.0
```

**Comment by tlambert03** [comment]
I'm sorry, I just tried this with python 3.7 and as similar of an environment as I could and couldn't repeat on mac.  my ubuntu setup is currently bonked so, will take me a while to rebuild that to test this out there.

**Comment by sofroniewn** [comment]
Cross linking to #2238 as it will change some of this behaviour

**Comment by jni** [comment]
I wonder if this is related to my random affinder window launch issue, which also could not be reproduced in a fresh env. @GenevieveBuckley could you try to create a minimal environment to see whether it also happens in a fresh env?

**Comment by tlambert03** [comment]
was able to try this on ubuntu 20.04 and saw the black screen.  will see what I can learn.  @jni, have you not seen this yet?

**Comment by tlambert03** [comment]
ok, so it looks like I can reproduce this (on ubuntu) without even using napari:

```python
from qtpy.QtWidgets import QApplication, QMainWindow
app = QApplication([])
mw = QMainWindow()
mw.show()  # black screen shows
app.exec_()  # it turns into the "standard" white window (like `napari.run()`)
```

need to figure out why this happens on ubuntu, and whether it's unique to 20.04.  @goanpeca, any insight here?

**Comment by jni** [comment]
> @jni, have you not seen this yet?

I haven't encountered this because `%gui=qt` is part of my IPython default profile. 😂

**Comment by tlambert03** [comment]
@GenevieveBuckley, can you try something for me?

in `qt_main_window.py`, in `Window.__init__` ... change this
```python
        if show:
            self.show()
```
to this:
```python
        if show:
            from qtpy.QtCore import QTimer
            QTimer.singleShot(0, self.show)
```

and see if it avoids the black screen until you use `napari.run`?

**Comment by GenevieveBuckley** [comment]
I've tried your suggestion @tlambert03 and it seems to make things worse, not better. Sorry!

```python
(napari-dev) genevieve@genevieve-G5-5500:~$ ipython
Python 3.7.9 | packaged by conda-forge | (default, Dec  9 2020, 21:08:20) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.20.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import napari

In [2]: viewer = napari.Viewer()

In [3]: viewer.show()  # I still see the black screen appear here

In [4]: viewer.run()  # but now things are broken instead of me seeing the normal napari window
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-3a674e6670be> in <module>
----> 1 viewer.run()

AttributeError: 'Viewer' object has no attribute 'run'

```

**Comment by tlambert03** [comment]
it should be `napari.run` ...  not `viewer.run`

but in any case, the window shouldn't have shown at ` viewer.show()`.  I'm really confused (and have no idea how linux is showing a widget that explicitly has not been shown without an event loop ... do this without napari if you can:

```python
from qtpy.QtWidgets import QApplication, QMainWindow
app = QApplication([])
mw = QMainWindow()

# do you see anything here?  If so, we're probably screwed.

# however... here's where I'd expect you to see the black screen
mw.show()

```
if you see the black screen there.  try this:

```python
from qtpy.QtWidgets import QApplication, QMainWindow
from qtpy.QtCore import QTimer

app = QApplication([])
mw = QMainWindow()
QTimer.singleShot(0, mw.show)  # do you see a black screen?

# if not ... then do this:
app.exec_()  # do you see a non-black screen?
```

**Comment by tlambert03** [comment]
I just got my linux system back up and running, and this fix _does_ work for me to prevent the black box.  @GenevieveBuckley, I'm not sure what went wrong when you tested it above, but maybe you can try #2276, which is what I was trying to suggest in https://github.com/napari/napari/issues/2233#issuecomment-777864183

**Comment by GenevieveBuckley** [comment]
> it should be `napari.run` ... not `viewer.run`

Gah, of course - sorry!

**Comment by GenevieveBuckley** [comment]
> do this without napari if you can:
> 
> ```python
> from qtpy.QtWidgets import QApplication, QMainWindow
> app = QApplication([])
> mw = QMainWindow()
> 
> # do you see anything here?  If so, we're probably screwed.
> 
> # however... here's where I'd expect you to see the black screen
> mw.show()
> ```
> 

We're not screwed, this works the way you expected & wrote above.

> if you see the black screen there. try this:
> 
> ```python
> from qtpy.QtWidgets import QApplication, QMainWindow
> from qtpy.QtCore import QTimer
> 
> app = QApplication([])
> mw = QMainWindow()
> QTimer.singleShot(0, mw.show)  # do you see a black screen?
> 
> # if not ... then do this:
> app.exec_()  # do you see a non-black screen?
> ```

Yep, this also works like you expect. I don't see any kind of screen until the last line, when `app.exec_()` pops up a grey dialog box

**Comment by GenevieveBuckley** [comment]
> Hi @GenevieveBuckley,
> 
> The reason for this is that it's no longer an error to _create_ a viewer without using gui_qt first (like it used to be). But you do need to use either gui_qt or `napari.run()` (the latter is preferable) to actually show the viewer. Think of this now as similar to having to use plt.show in matplotlib.
> 
> Basically, we're phasing out gui_qt. See [#2056 (comment)](https://github.com/napari/napari/pull/2056#discussion_r550924887) for more background and rationale on this new behavior. And let us know what you think

I apologize, I totally missed this earlier.

I think:
* It's fantastic you're doing this work making sure headless napari can work well
* I worry that for new users this might become a bit of a trap, and we don't have good ways to redirect them back to what they thought they were doing in the first place. It's only "not a mistake" if you *meant* to be running napari headless. If you didn't mean that, this is a very fast road for newcomers to "napari is a piece of crap, it doesn't even work for the intro tutorials".

**Comment by GenevieveBuckley** [comment]
Since the removal of the user warning was intended, I'll close this issue.

I would like it if we had a better way to address this barrier to adoption:

> I worry that for new users this might become a bit of a trap, and we don't have good ways to redirect them back to what they thought they were doing in the first place. It's only "not a mistake" if you meant to be running napari headless. If you didn't mean that, this is a very fast road for newcomers to "napari is a piece of crap, it doesn't even work for the intro tutorials".

**Comment by tlambert03** [comment]
>  It's only "not a mistake" if you meant to be running napari headless.

would you then suggest that it's a mistake to make a matplotlib figure without showing it?

> "napari is a piece of crap, it doesn't even work for the intro tutorials".

I'm not sure I agree on that ... surely, parity between the tutorials and the current state of master is an independent topic from how things "should work" ... for sure, the tutorials need to be clear and up-to-date,  but we have to be able to evolve from how the tutorials were originally drafted, and yes, we'll need to update our tutorials from time to time.  (I'd be all for completing removing `napari.gui_qt` from our tutorials and moving to the new syntax!)
with that said ... all of the tutorials still work!  `%gui qt` works and `napari.gui_qt` works... we never _suggest_ in our tutorials that you should use napari without one of those.

basically, the only "fix" to your request that I can think of here is that we autodetect whether someone is using ipython without having %gui qt active, and suggest to them that they might "like to try" %gui qt.  As I see it, that's a _optional_ third party feature that napari happens to work well with, but it should not be an error (or even a warning) if you don't use it.

**Comment by tlambert03** [comment]
>  if you meant to be running napari headless. 

it's also not even about running strictly headless or not.  If you're _not_ using ipython's magic `%gui qt`, (which again, is an optional third party event loop feature), then you would definitely need to set up some additional things before starting the event loop with `napari.run()`.  That's why we had the `gui_qt` context in the first place (because no one should _have_ to use %gui qt)... but for all the reasons mentioned in https://github.com/napari/napari/pull/2056#discussion_r550924887, it was never well-suited to a context manager... so the preferred approach, if you're not using %gui qt, is now like matplotlib:

```python
viewer = napari.Viewer()
# hook up all stuff that you'd like to do before you actually "run" the viewer
napari.run()
```

**Comment by jni** [comment]
> basically, the only "fix" to your request that I can think of here is that we autodetect whether someone is using ipython without having %gui qt active, and suggest to them that they might "like to try" %gui qt. As I see it, that's a optional third party feature that napari happens to work well with, but it should not be an error (or even a warning) if you don't use it.

This is the crux of it and I wonder if it's not a good idea to do this anyway? Especially given the large community we've built used to `gui_qt()`/`%gui qt`. ie *if* we're in IPython *and* `%gui qt` hasn't been launched when we instantiate the viewer, *warn* the user that they want to use `napari.run()` or `%gui qt`. What do you think @tlambert03?

**Comment by tlambert03** [comment]
as i laid out above, I'm not a fan... but it's a democracy and I'm happy to be outvoted! :joy:
I absolutely think we have an obligation to those who _already_ use `gui_qt()` to make sure it doesn't break without a solid deprecation period.  And I also definitely think we should continue to tell people about `%gui qt` in our documentation, just as we always have, and to make sure that it works well with napari.   But I use napari in IPython all the time without `%gui qt` and I really don't want to be warned about it.  It feels akin to matplotlib warning you everytime you made a figure "hey you forgot to use `plt.ion()`!"

**Comment by jni** [comment]
> But I use napari in IPython all the time without %gui qt

You monster! 😂

**Comment by GenevieveBuckley** [comment]
I think it's a bit different to the matplotlib example because in my view the real power with napari is interactivity, whereas in matplotlib you generally want to construct the entire figure before seeing it.

But I'm not trying to start a philosophical argument here. I just missed that there was an intentional change here & was trying to be  helpful by flagging it.

**Comment by tlambert03** [comment]
I think it's the warning for completely acceptable usage that bothers me.  I feel like the "use `%gui qt`" suggestion belongs in the documentation and tutorials more than the code itself.   However, an alternative that I would be less opposed to is actually _making_  the `napari.Viewer` be interactive by default when running in IPython.  This could be done as follows:

```python
# qt_main_window.py
class Window:
    ...

    def show(self, interactive=True):
        from .qt_event_loop import _ipython_has_eventloop
        from ..utils.misc import in_ipython

        if interactive and in_ipython() and not _ipython_has_eventloop():
            from IPython import get_ipython

            get_ipython().enable_gui('qt')
        ...
```

_edit:_ I'll note here that starting `napari.Viewer()` in this case will switch focus over to the new window... which @jni might not like, but I believe that's the same loss of focus you normally get (or at least _I_ get), when running `%gui qt`

**Comment by GenevieveBuckley** [comment]
> However, an alternative that I would be less opposed to is actually _making_ the `napari.Viewer` be interactive by default when running in IPython. 

I'd support this, on the basis that headless mode is likely to be less commonly used (and when it is used, those power users are likely to be people with stronger coding skills & ability to troubleshoot).
